turnstile-example/turnstile/examples/cs478/current/stlc.rkt

```diff
@@ -442,3 +445,43 @@
   #;  [⊢ (if- (inl-? e-)
           (let- ([x- (inl--val e-)]) e1-)
           (let- ([y- (inr--val e-)]) e2-)) ⇒ τ])
+
+(provide typed-match)
+(define-typerule typed-match
+  [(_ ([x:id e] [pat e-next] ...) b) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   [[x ≫ x- : τ] ⊢ (typed-match ([pat e-next] ...) b) ≫ rest- ⇒ τ-out]
+   ---------------------------------
+   [⊢ (let- ([x- e-]) rest-) ⇒ τ-out]]
+  [(_ ([[x:id ...] e] [pat e-next] ...) b) ≫
+   [⊢ e ≫ e- ⇒ (~× τ-tup ... ~!)]
+   #:fail-unless (= (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   (format "~a pattern variables does not match ~a values in tuple" (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   [[x ≫ x- : τ-tup] ... ⊢ (typed-match ([pat e-next] ...) b) ≫ rest- ⇒ τ-out]
+   -----------------------------------------
+   [⊢ (match- e- [(list x- ...) rest-]) ⇒ τ-out]]
+  [(_ ([(fs:flds) (~and e (~parse (rec [l (~datum =) e-val] ...) #'e))] [pat e-next] ...) b) ≫
```

Comment:
I think you don't need the `~and` here, because you are matching on the exact syntax.
It's only needed for the type below because we need to invoke the `~Rec` pattern macro

Reply:
You may need `~literal` around `rec` though